core/shared/src/main/scala/zio/ZIOConcurrent.scala

```diff
@@ -0,0 +1,153 @@
+package zio
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.ConcurrentLinkedQueue
+import scala.annotation.tailrec
+import zio.Synchronizer._
+
+final class CountDownLatch(count: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(count, maxWaiters)
+
+  // TODO is it possible to evaluate once and reuse the Promise?
+  private[this] val awaitZero = sync.acquire(count)
```

Comment:
`awaitZero` is an `UIO` so I imagine this might not work because the promise will be created on every evaluation?

Reply:
I think we would want the creation of the `CountDownLatch` to be an effect (e..g `def make: UIO[CountDownLatch]`) because it allocates mutable state and then we could create the `Promise` as part of creating the `CountDownLatch`.